src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Public/EditorContextScope.cs

```diff
@@ -228,5 +282,188 @@ public Fx2NLParameters GetFx2NLParameters()
                  UsageHints = this.UsageHints
             };
         }
+
+        public DefinitionsCheckResult CheckUserDefinedFunctions(string expression)
+        {
+            return _checkUserDefinedFunctions?.Invoke(expression) ?? throw new NotImplementedException();
+        }
+
+        public IIntellisenseResult Suggest(string expression, int cursorPosition, LSPExpressionMode mode)
+        {
+            return mode switch
+            {
+                LSPExpressionMode.Default => ((IPowerFxScope)this).Suggest(expression, cursorPosition),
+                LSPExpressionMode.UserDefinedFunction => SuggestForUdf(expression, cursorPosition),
+                _ => throw new ArgumentException($"Unknown LSP mode {mode}")
+            };
+        }
+
+        private IIntellisenseResult SuggestForUdf(string expression, int cursorPosition)
+        {
+            var checkUdf = CheckUserDefinedFunctions(expression);
+            var parsedUdfs = checkUdf.ApplyParse();
+
+            // Add valid functions to engine if not already done.
+            var funcs = checkUdf.ApplyCreateUserDefinedFunctions();
+            _localUDFs = new SymbolTable();
+            _localUDFs.AddFunctions(funcs);
+
+            // $$$ find a better way to get engine.
```

Comment:
> // $$$ find a better way to get engine [](http://example.com/codeflow?start=11&length=39)

is this ok?

Reply:
not ideal, but pragmatic enough. Not really a better way.